### DISCUSSION TITLE
Help with clickhouse-compressor

### DISCUSSION THREAD

**Discussion by DouglasPersson** [ClickHouse/ClickHouse#47262] [Q&A] [answered]
Hello!
We are trying to create an automated way of selecting the optimal compression algorithm for a given column in ClickHouse in our masters thesis. We are trying to analyse the results from **clickhouse-compressor** using different combinations of codecs and compression algorithms to get _compression ratio_, _compression time_, _decompression time_ etc.

However, we run into multiple errors when setting the codecs (_Delta_, _DoubleDelta_ and _Gorilla_). Here are some examples:

**1.**
**Command:**
`clickhouse-compressor --codec 'Delta' --codec 'LZ4' -- input '<INPUT_FILE>' --output '<OUTPUT_FILE>'`
**Error:**
`floating point exception`

**2.**
**Command:**
`clickhouse-compressor --codec 'Delta(8)' --codec 'LZ4' -- input '<INPUT_FILE>' --output '<OUTPUT_FILE>'`
**Error:**
No error, this works!
(Works for all supported numbers: 1, 2, 4 and 8).

**3.**
**Command:**
`clickhouse-compressor --codec 'DoubleDelta' --codec 'LZ4' -- input '<INPUT_FILE>' --output '<OUTPUT_FILE>'`
**Error:**
`segmentation fault`

**3.**
**Command:**
`clickhouse-compressor --codec 'Gorilla' --codec 'LZ4' -- input '<INPUT_FILE>' --output '<OUTPUT_FILE>'`
**Error:**
`floating point exception`

We have tried running the clickhouse-compressor on both actual part files, exported csv files as well as exported native files. We have tried with Nullable(UInt64) (with and without null values) as well as Nullable(Float64) (also, with and without null values) and we get the same result every time.

Does anyone know to get the clickhouse-compressor to work with all codecs and compression algorithms? We would like to get the compression and decompression times with as little overhead as possible.

Thanks in advance!

**Comment by mdonkers** [comment, marked as answer]
I think this question was asked in one of our other channels as well. So just copying the response here for others to find;

(from @Avogar )

> As I can see from the code, `DoubleDelta`, `Delta` (without any arguments) and `Gorilla` codecs takes bytes size from column type. And if there is no column type provided (what we see in `clickhouse-compressor`), we set bytes size to 0, that leads to division by zero error. By now I guess it’s impossible to use such codecs with `clickhouse-compressor`. But we can fix it easily (for example, allow arguments for `DoubleDelta`  and `Gorilla` codecs to specify byte size explicitly) and also we should throw an exception when bytes size is 0.

PR that was created to fix these issues: https://github.com/ClickHouse/ClickHouse/pull/47271

**Reply by DouglasPersson** [reply]
Thanks for the response! We have tested the fix and everything seems to work as expected!

**Comment by UnamedRus** [comment]
TBH, 
table sorting key impact compression much more than compression codecs.

https://github.com/ClickHouse/ClickHouse/issues/18536

**Reply by DouglasPersson** [reply]
Yes we are aware of that, but unfortunately that is outside the scope of our thesis. However, we will definitely suggest it for future work. Thanks for sharing!
